File: lib/serialization.js
module.exports = function(lib, xelib, helpers) {
    let { GetString, Fail } = helpers;

    // SERIALIZATION METHODS
    Object.assign(xelib, {
        ElementToJSON: function(id) {
            return GetString(function(_len) {
                if (!lib.ElementToJson(id, _len))
                    Fail(`Failed to serialize element to JSON: ${id}`);
            });
        },
        ElementToObject: function(id) {
            return JSON.parse(this.ElementToJSON(id));
        }
    });
}

File: src/cpp/serialization.cc
#include <nan.h>
#include "serialization.h"
#include "lib.h"
#include "helpers.h"

using namespace Nan;

NAN_METHOD(ElementToJson) {
    Cardinal handle = GetCardinal(info[0]);
    PInteger len = (PInteger) BufferToPointer(info[1]);
    WordBool success = xelib.functions.ElementToJson(handle, len);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(ElementFromJson) {
    Cardinal handle = GetCardinal(info[0]);
    PWChar path = (PWChar) BufferToPointer(info[1]);
    PWChar json = (PWChar) BufferToPointer(info[2]);
    WordBool success = xelib.functions.ElementFromJson(handle, path, json);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(DefToJson) {
    Cardinal handle = GetCardinal(info[0]);
    PInteger len = (PInteger) BufferToPointer(info[1]);
    WordBool success = xelib.functions.DefToJson(handle, len);
    info.GetReturnValue().Set(Nan::New((bool) success));
}